### PR TITLE
arch: arm64: mmu: Restore SRAM region

### PR DESCRIPTION
**Pull request by carlocaione**
In a5f34d85c2e3 ("soc: arm: qemu_cortex_a53: Remove SRAM region") the
SRAM memory region was removed.

While this is correct when userspace is not enabled, when userspace is
enabled new regions are introduced outside the boundaries of
the mapped [__kernel_ram_start,__kernel_ram_end] region. This means that
we need to map again the whole SRAM to include all the needed regions.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

### CODE REVIEW

File: arch/arm/core/aarch64/arm_mmu.c [hunk 1]
```diff
@@ -328,6 +328,10 @@ static void init_xlat_tables(const struct arm_mmu_region *region)
 
 /* zephyr execution regions with appropriate attributes */
 static const struct arm_mmu_region mmu_zephyr_regions[] = {
+	MMU_REGION_FLAT_ENTRY("SRAM",
+			      (uintptr_t)CONFIG_SRAM_BASE_ADDRESS,
```
Comment (ioannisg) [inline arch/arm/core/aarch64/arm_mmu.c:332]: @carlocaione is it possible to use here linker symbols?
In the same way the regions, below are defined.
I think this would be consistent, at least with cortex-m

Comment (ioannisg) [reply arch/arm/core/aarch64/arm_mmu.c:332]: E.g. _image_ram_start, or whatever is available for ARM_64....

Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:332]: Yes it is possible. I don't have a strong preference here but I honestly thought that using `CONFIG_*` symbols was better than using something exported by the linker.

Comment (ioannisg) [reply arch/arm/core/aarch64/arm_mmu.c:332]: It is not the same. The linker symbols mark the SRAM that is used by the Zephyr image. The CONFIG_ variables map to physical SRAM available in the SoC. These might not match if, say, we're using SRAM partitioning. It is safer to use linker symbols.

That's what is also done right below, with text and RO data areas

**Comment by ioannisg** [comment]
@Abhishek-brcm @sandeepbrcm please, review, as well

File: arch/arm/core/aarch64/arm_mmu.c (hunk 1 above)
Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:332]: Using `_image_ram_start` with size `_image_ram_end - _image_ram_start` is not going to work because we need also to map the remaining space between `_image_ram_end` and `_kernel_ram_end` (that in this case is coincident with the end of the SRAM).

So we should map from `_image_ram_start` up to `__kernel_ram_end` as you did on Cortex-M. But that is basically the same as mapping between `CONFIG_SRAM_BASE_ADDRESS` with size `CONFIG_SRAM_SIZE` according to the content of `include/arch/arm/aarch64/scripts/linker.ld` (excluding XIP & co).

For sure I'm missing something here.

Comment (ioannisg) [reply arch/arm/core/aarch64/arm_mmu.c:332]: I am just saying it is more robust using linker symbols because you are always in control of your image. Using Kconfig symbols is fragile because we may change what SRAM_BASE_ADDRESS and SRAM_SIZE mean, actually.

Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:332]: I mean, according to this reasoning we should avoid to use `CONFIG_*` symbols all the time in our code because we may change what they mean mid-development? Without considering that there is other code using these symbols including X86 code and MMU code. 

To be clear it's not matter of principle here but since there is no linker script symbol equivalent of `SRAM_SIZE` and I cannot compute it at run/load-time from the linker symbols I have already (because I have a static array of structs for the memory regions) I have to either change how the SRAM region is allocated into the driver or add a new linker symbol myself that I don't think it's much better than using `CONFIG_*`.

Comment (andrewboie) [reply arch/arm/core/aarch64/arm_mmu.c:332]: > we may change what SRAM_BASE_ADDRESS and SRAM_SIZE mean, actually.

I think this is turning into a bikeshed.
There's no equivalent linker symbols for the entire RAM bounds.
We also have a fair amount of code which is already using CONFIG_SRAM_BASE_ADDRESS and CONFIG_SRAM_SIZE.
Where are we are getting the impression that these Kconfigs are going to change?

I think if we want to get rid of these defines in place of something else, file a bug on it.

**Review by tyrantlizard77 — APPROVED** [review]
I don't have any objections to this particular change, but sooner rather than later I think you need to get rid of these static regions in favor of an arch_mem_map() implementation.

**Comment by carlocaione** [comment]
> I don't have any objections to this particular change, but sooner rather than later I think you need to get rid of these static regions in favor of an arch_mem_map() implementation.

Yes, I agree. That's already in the TODO list but a bit down on the list right now. But I'm fully aware of that, thanks.

File: arch/arm/core/aarch64/arm_mmu.c [hunk 2]
```diff
@@ -340,15 +344,6 @@ static const struct arm_mmu_region mmu_zephyr_regions[] = {
...
 			      MT_RODATA | MT_SECURE),
-
-	/* Mark rest of the zephyr execution regions (data, bss, noinit, etc.)
-	 * cacheable, read-write
-	 * Note: read-write region is marked execute-ever internally
-	 */
-	MMU_REGION_FLAT_ENTRY("zephyr_data",
-			      (uintptr_t)__kernel_ram_start,
-			      (uintptr_t)__kernel_ram_size,
-			      MT_NORMAL | MT_RW | MT_SECURE),
```
Comment (Abhishek-brcm) [inline arch/arm/core/aarch64/arm_mmu.c:351]: Basic idea here is to mark the **zephyr kernel regions**, such as text, rodata and data sections appropriately. And yes, these come from linker symbols. We do not want to mark the entire memory regions (such as SRAM) in here.

We removed SRAM marking from qemu_cortex_a53 as it was indicated that the entire SRAM is covered through regions marked here. If that is not the case anymore, please revert and put SRAM region back into qemu_cortex_a53 file rather than doing this.

Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:351]: > Basic idea here is to mark the **zephyr kernel regions**, such as text, rodata and data sections appropriately. And yes, these come from linker symbols. We do not want to mark the entire memory regions (such as SRAM) in here.

We want to create an identity mapping for the whole SRAM and for that we do not have the linker symbols, so we are using the `CONFIG_*` symbols. We want to do this because this is a preliminary work to move to user-mode and user threads that are mapped outside the zephyr kernel regions.

> We removed SRAM marking from qemu_cortex_a53 as it was indicated that the entire SRAM is covered through regions marked here. If that is not the case anymore, please revert and put SRAM region back into qemu_cortex_a53 file rather than doing this.

Why? This is something that we always want, not something that should be SoC-specific.

Comment (Abhishek-brcm) [reply arch/arm/core/aarch64/arm_mmu.c:351]: > > Basic idea here is to mark the **zephyr kernel regions**, such as text, rodata and data sections appropriately. And yes, these come from linker symbols. We do not want to mark the entire memory regions (such as SRAM) in here.
> 
> We want to create an identity mapping for the whole SRAM and for that we do not have the linker symbols, so we are using the `CONFIG_*` symbols. We want to do this because this is a preliminary work to move to user-mode and user threads that are mapped outside the zephyr kernel regions.
> 
No objection to mapping the whole SRAM or using the `CONFIG_*` symbols. I was just pointing out that purpose of marking zephyr kernel region is being satisfied by linker symbols.

> > We removed SRAM marking from qemu_cortex_a53 as it was indicated that the entire SRAM is covered through regions marked here. If that is not the case anymore, please revert and put SRAM region back into qemu_cortex_a53 file rather than doing this.
> 
> Why? This is something that we always want, not something that should be SoC-specific.

yes, I guess that's right, SRAM (CONFIG_SRAM_BASE) is always used for all SoCs per current linker script. However, data section may/may not be part of SRAM if custom linker script is used, so should not be removed.

Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:351]: > yes, I guess that's right, SRAM (CONFIG_SRAM_BASE) is always used for all SoCs per current linker script. However, data section may/may not be part of SRAM if custom linker script is used, so should not be removed.

But if you have a custom linker script shouldn't then the mapping of the data section be done in the board / platform specific file?

Comment (Abhishek-brcm) [reply arch/arm/core/aarch64/arm_mmu.c:351]: No.. marking of zephyr kernel regions is abstracted from user.

Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:351]: That's not my point. If you have the `zephyr_data` region mapped outside the SRAM in some custom place (for example somewhere in DRAM) then that is something that you should specify in a MMU region defined in the board / soc file. In your case should be somewhere in `soc/arm/bcm_vk/viper/mmu_regions.c`. In the standard case the `zephyr_data` region is inside the SRAM with the same flags / permissions, I don't think why we should specify that as a standalone region.

Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:351]: @sandeepbrcm I'm not exactly sure what you are suggesting here because the SRAM is defined as a separate entry in this patch.

Bottom line is that we need to have the SRAM region mapped and we cannot have a platform without `CONFIG_SRAM*` so this is something that we should have in the core driver. Otherwise we force every single platform to have a MMU region to map the SRAM, this is nonsense. FWIW this is following what's being already done by X86 (see https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/x86/gen_mmu.py).

@andrewboie @ioannisg any take  on this?

Comment (sandeepbrcm) [reply arch/arm/core/aarch64/arm_mmu.c:351]: I am suggesting to try
1- Add this entry for SRAM , because we know its an SRAM from the defines.
2- Retain   `zephyr_data `mapping using ` __kernel_ram_start`,  `__kernel_ram_size`  So that zephyr OS does not rely on any configs but map its own regions from linker symbols.

Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:351]: But why considering that in the standard case`zephyr_data` region is a subset of the SRAM region with the same attributes / permissions?

Comment (sandeepbrcm) [reply arch/arm/core/aarch64/arm_mmu.c:351]: - map SRAM ( default attributes)
- map rest of the sections with their attributes.  zephyr_data section happens to be same as default

 It is only making this `zephyr_data`  section entry explicit in code, like a documentation without any assumption of SRAM map. Seems harmless . After all the final table entries are same either case.

Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:351]: It depends on your definition of harmless. The way the code builds the page tables is not smart enough to detect that the two memory regions are overlapping so the whole `zephyr_data` section is mapped again in a different table and it even creates an additional table to host the duplicate entries.

Comment (Abhishek-brcm) [reply arch/arm/core/aarch64/arm_mmu.c:351]: > the whole `zephyr_data` section is mapped again in a different table and it even creates an additional table to host the duplicate entries.

Let me correct you here. Additional table is not created for `zephyr_data` section. If SRAM has been entirely mapped and, `zephyr_data` section is mapped again; considering it is part of SRAM, logic will just do the mapping at the corresponding PTEs and PTEs will effectively remain unchanged as both SRAM and `zephyr_data` section have same attributes.

Now assume, if `zephyr_data` section is not in SRAM, then user does not need to worry about mapping, zephyr itself will take care of it, precisely why we are requesting to keep `zephyr_data` section entry explicit in code.

Comment (carlocaione) [reply arch/arm/core/aarch64/arm_mmu.c:351]: > Let me correct you here. Additional table is not created for `zephyr_data` section. If SRAM has been entirely mapped and, `zephyr_data` section is mapped again; considering it is part of SRAM, logic will just do the mapping at the corresponding PTEs and PTEs will effectively remain unchanged as both SRAM and `zephyr_data` section have same attributes.

Are you sure about this? Because enabling the debugging it looks like this is not the case.

> Now assume, if `zephyr_data` section is not in SRAM, then user does not need to worry about mapping, zephyr itself will take care of it, precisely why we are requesting to keep `zephyr_data` section entry explicit in code.

I think we are bikeshedding again honestly. I'll add back the `zephyr_data` back so we can move to some more beefy PR about MMU ;)

Comment (Abhishek-brcm) [reply arch/arm/core/aarch64/arm_mmu.c:351]: > > Let me correct you here. Additional table is not created for `zephyr_data` section. If SRAM has been entirely mapped and, `zephyr_data` section is mapped again; considering it is part of SRAM, logic will just do the mapping at the corresponding PTEs and PTEs will effectively remain unchanged as both SRAM and `zephyr_data` section have same attributes.
> 
> Are you sure about this? Because enabling the debugging it looks like this is not the case.
> 
yes, I am sure. Please check PTE address for both, which part of `MMU_DEBUG` prints (to see PTE address, please set  `DUMP_PTE` to 1).

> > Now assume, if `zephyr_data` section is not in SRAM, then user does not need to worry about mapping, zephyr itself will take care of it, precisely why we are requesting to keep `zephyr_data` section entry explicit in code.
> 
> I think we are bikeshedding again honestly. I'll add back the `zephyr_data` region so we can move to some more beefy PR about MMU ;)

important  to get basics right :)

Comment (andrewboie) [reply arch/arm/core/aarch64/arm_mmu.c:351]: I can speak to the intent behind the x86 code.

What we produce on x86 is a build time set of page tables with all RAM addresses identity mapped. This is so these page tables can be installed at early boot and the kernel can continue execution as normal. On x86_64, you can't enter 64-bit mode without page tables.

What's really needed at minimum is to have all the RAM used by the kernel at boot, which in turn is defined by where things get linked. We need all program text, read-only data, and data mapped to where they are linked. If symbols are linked in a RAM region, we will need a mapping for it. The kernel assumes no runtime memory mappings are necessary for any RAM that is part of the linked kernel image once we get into z_cstart().

In addition, if `CONFIG_SRAM_REGION_PERMISSIONS` is turned on, then program text, read-only data, and data are in their own page-aligned regions and have read/execute/user, read/user, and read/write permissions respectively.

We might need some special handling if we have a physical memory map with sparse RAM regions. For example, the kernel assumes that virtual memory mappings may start at virtual address (CONFIG_SRAM_BASE_ADDRESS + KB(CONFIG_SRAM_SIZE)) up to (CONFIG_SRAM_BASE_ADDRESS + CONFIG_KERNEL_VM_SIZE). But if there is some other RAM region we need to idenity-map that overlaps this, we could get into trouble.

However, there is no requirement that RAM is identity-mapped. It just needs to all be mapped in the address space, and a defined region for additional memory mappings. The kernel will work fine if you link it virtually. So if you have multiple RAM regions, it may be better to just link the kernel at some fixed virtual address, and have all the RAM mappings, discontinuous physically, be one continuous region in virtual memory. For most systems identity-mapped RAM is just fine but you aren't forced to do it. The Kconfigs used by kernel/mmu.c may need some revision however.

If you don't identity map ram your arch code will be more complex as will be your linker script. I have a branch here on x86 where I was experimenting with it: https://github.com/andrewboie/zephyr/tree/vm-linking I also used a technique where the kernel was double-linked in the page tables at both its virtual and physical base addresses to transition the instruction pointer to the virtual domain at early boot. If this is of interest I can try to rebase this work.

x86 currently also identity maps *all* of system RAM in these page tables. This includes "unused" pages, past the end of the kernel image, to the RAM boundary. These "unused" pages are currently used to create the newlib heap in a way that doesn't involve memory mappings, but this is expected to change soon, along with other infrastructure for anonymous memory mappings and a core kernel ontology of physical page frames. But even with this, we'll probably still keep the identity mapping for these pages (they would be mapped twice, or re-mapped) so I don't expect this to change things much from your perspective.

HTH
Andrew

Comment (andrewboie) [reply arch/arm/core/aarch64/arm_mmu.c:351]: the tl;dr version: if multiple physical RAM regions exist I think we need to consider not identity-mapping RAM and instead mapping it all in one virtual region, with linkage at these regions. I can dig up some prototyping I did for this on x86, I had this working well on 32-bit.

Comment (andrewboie) [reply arch/arm/core/aarch64/arm_mmu.c:351]: https://github.com/zephyrproject-rtos/zephyr/pull/29803 should allow this on the core kernel side.

File: arch/arm/core/aarch64/mmu/arm_mmu.c [hunk 3]
```diff
@@ -272,15 +287,6 @@ static const struct arm_mmu_region mmu_zephyr_regions[] = {
 			      (uintptr_t)_image_rodata_start,
 			      (uintptr_t)_image_rodata_size,
 			      MT_NORMAL | MT_P_RO_U_NA | MT_SECURE),
-
-	/* Mark rest of the zephyr execution regions (data, bss, noinit, etc.)
```
Comment (Abhishek-brcm) [inline arch/arm/core/aarch64/mmu/arm_mmu.c:276]: unnecessary diff generated.

Comment (carlocaione) [reply arch/arm/core/aarch64/mmu/arm_mmu.c:276]: I purposely moved `zephyr_data` close to the SRAM region as a reminder that the two regions are mostly overlapping.

Comment (Abhishek-brcm) [reply arch/arm/core/aarch64/mmu/arm_mmu.c:276]: ok

**Review by Abhishek-brcm — APPROVED** [review]

**Review by sandeepbrcm — APPROVED** [review]
